---
title: "2020005_hypothalamus-circadian-investigation-rna-seq_Zhang"
author: "Jiayu Zhang"
date: "4/5/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = FALSE)
```

# Hypothalamus Investigation Group Comparison  

> general data structure  
> differential expressed genes (control group vs. exercise group) t-test?
> orthologs of mouse circadian genes in rats
> subset the differential expressed genes -- circadian and exercise

### set up the environment
```{r}
################################################################################
##### Resources and Dependencies ###############################################
################################################################################

# Set the working directory
WD <- '/Volumes/SanDisk64GB/MoTrPAC/'
setwd(WD)

# Load dependencies
pacs...man <- c("tidyverse","GenomicRanges", "DESeq2","devtools","rafalib","GO.db","vsn","hexbin","ggplot2",
               "GenomicFeatures","Biostrings","BSgenome","AnnotationHub","plyr","dplyr",
               "org.Rn.eg.db","pheatmap","sva","formula.tools","pathview","biomaRt","feather",
               "PROPER","SeqGSEA",'purrr','BioInstaller','RColorBrewer','lubridate', "hms","ggpubr", "ggrepel",
               "AnnotationDbi")

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("PROPER")
BiocManager::install("org.Rn.eg.db")
BiocManager::install("pcaExplorer")
BiocManager::install("GenomicFeatures")
BiocManager::install("S4Vector")
BiocManager::install("qvalue")
mypackages = c("tidyverse","GenomicRanges", "DESeq2","devtools","ggplot2","dplyr","PROPER", "lubridate", "hms",
               "org.Rn.eg.db", "AnnotationDbi", "pcaExplorer", "pheatmap","GenomicFeatures", "genefilter", "qvalue")
lapply(mypackages, FUN = function(X) {
  do.call("library", list(X)) })

############################################################
##### Functions ############################################
############################################################

# Set select
select <- dplyr::select

# Capture the Date and Author
################################################################################
date <- format.Date( Sys.Date(), '%Y%m%d' )
auth <- "Jiayu"
################################################################################
```

## Data structure  
### Load and clean data

* Data file to load:
  + Standford batch 2 & metadata & phenotype 
```{r}
# Stanford batch 2 read counts (320 samples in total; 80 hypothalamus samples)
sford_2 <- read.table(file = paste0(WD,'Data/rna-seq/stanford/batch_2/summary/rsem_genes_count.txt'), 
                      header = T, row.names = 1, na.strings = "NA", check.names = F)  %>% as.matrix()

# Stanford batch 2 meta data
sford_meta_2 <-  read.csv(file = paste0(WD,'Data/rna-seq/stanford/batch_2/metadata/sample_metadata_20191010.csv'), 
                          header = T, na.strings = "NA", check.names = F) 

# Merged Phenotype
###################################

load(paste0(WD,"Data/phenotype/merged/merged_dmaqc_data2019-10-13.Rdata"))

# Mice RNA seq library
###################################

# circadian genes
# TODO: may need to find orthologs for rats
mice_cir_genes = read.csv(file = paste0(WD,'Data/library/20080516_mouse-SCN-circadian-genes_yan.csv'), 
                            header = T, na.strings = "NA", check.names = F)
mice_cir_genes = mice_cir_genes[-which(is.na(mice_cir_genes$SCN)),]

# Convert counts to integers
mode(sford_2) <- "integer"

# Adjust column objects
########################
# To factors

sford_meta_2$vial_label = as.factor(sford_meta_2$vial_label)
factor_cols <- c("labelid",
                 "viallabel",
                 "animal.registration.sex")
for(fc in factor_cols){
  merged_dmaqc_data[[fc]] <- as.factor(merged_dmaqc_data[[fc]])
}

# To Dates: 03JUL2018
date_cols <- c("acute.test.d_visit",
               "acute.test.d_start",
               "animal.familiarization.d_visit",
               "animal.familiarization.d_treadmillbegin",
               "animal.familiarization.d_treadmillcomplete",
               "animal.registration.d_visit",
               "animal.registration.d_arrive",
               "animal.registration.d_reverselight",
               "specimen.collection.d_visit")
for(dc in date_cols){
  merged_dmaqc_data[[dc]] <- dmy(merged_dmaqc_data[[dc]])
}
# To Dates: Jan 2018 
merged_dmaqc_data <- merged_dmaqc_data %>%
  mutate(animal.registration.d_birth = dmy(paste0('1 ',animal.registration.d_birth)))

# To Times: 10:30:00
time_cols <- c("acute.test.t_complete",
               "specimen.collection.t_anesthesia",
               "specimen.collection.t_bloodstart",
               "specimen.collection.t_bloodstop",
               "specimen.collection.t_edtafill",
               "specimen.collection.uteruscomplete",
               "specimen.collection.t_uterusstart",
               "specimen.collection.t_uterusstop",
               "specimen.collection.t_death",
               "specimen.processing.t_collection",
               "specimen.processing.t_edtaspin",
               "specimen.processing.t_freeze")
for(tc in time_cols){
  merged_dmaqc_data[[tc]] <- merged_dmaqc_data[[tc]] %>% as.character() %>% parse_time()
}

# To Times: 01:21
merged_dmaqc_data <- merged_dmaqc_data %>%
  mutate(acute.test.howlongshock = 
           ifelse((merged_dmaqc_data$acute.test.howlongshock != ''), paste0('00:',acute.test.howlongshock), '00:00:00'))
merged_dmaqc_data$acute.test.howlongshock <- parse_time(merged_dmaqc_data$acute.test.howlongshock)

# Adjust labels within columns (e.g. 1 and 2 to female and male)
# see ../phenotype/merged/20191013_merged-column-dictionary_steep-edits.xlsx
########################
merged_dmaqc_data <- merged_dmaqc_data %>%
  mutate(animal.registration.sex = 
           case_when(animal.registration.sex == '1' ~ 'Female',
                     animal.registration.sex == '2' ~ 'Male'))
merged_dmaqc_data$animal.registration.sex <- as.factor(merged_dmaqc_data$animal.registration.sex)

#' ### Incorporate Annotation from Phenotypic Data
#' 

################################################################################
######### Incorporate Phenotypic Annotations ###################################
################################################################################

# Replace the name of the vial label column (for sake of join)
names(merged_dmaqc_data) <- str_replace(names(merged_dmaqc_data), 'viallabel', 'vial_label')

# Perform a left join to incorporate phenotypic information
status <- left_join(sford_meta_2, merged_dmaqc_data, by = 'vial_label') %>% as.data.frame()
#dim(status)

# Adjust objects in columns as needed

# Create new columns
status <- status %>% 
  mutate(Seq_batch = 
           case_when((GET_site == 'Stanford' & Seq_date == '190426') ~ "Stanford_1",
                     (GET_site == 'Stanford' & Seq_date == '190703') ~ "Stanford_2",
                     (GET_site == 'MSSM' & Seq_date == '190409') ~ "MSSM_1",
                     (GET_site == 'MSSM' & Seq_date == '190626') ~ "MSSM_2",
                     (GET_site == 'MSSM' & Seq_date == '190723') ~ "MSSM_3")
  )
#table(status$Seq_batch)

# To factors
factor_cols <- "Seq_batch"
for(fc in factor_cols){
  status[[fc]] <- as.factor(status[[fc]])
}

# "It is absolutely critical that the columns of the count matrix and the rows of the 
# column data (information about samples) are in the same order. DESeq2 will not make 
# guesses as to which column of the count matrix belongs to which row of the column data, 
# these must be provided to DESeq2 already in consistent order." ~ Mike Love

# TODO: Come back and figure out about missing samples and annotation

# Make sure that all values in status "sample_key" are unique
#status$sample_key %>% length()
#status$sample_key %>% unique %>% length()
# Make sure that all columns in counts matrix are unique
#colnames(all.data.m) %>% length()
#colnames(all.data.m) %>% unique %>% length()

rownames(status) <- status$vial_label
#all(rownames(status) %in% colnames(all.data.m))
#all(colnames(all.data.m) %in% rownames(status))
#all(rownames(status) == colnames(all.data.m))
#all.data.m <- all.data.m[, rownames(status)]
status$animal.key.anirandgroup = as.factor(status$animal.key.anirandgroup)
# add a new column to status dataframe to differentiate time in hour in a day
status$hour_death = hour(status$specimen.collection.t_death) %>% as.factor()
table(status$hour_death)
#hist(as.numeric(status$hour_death))
```

### Genomics Ranges (genel list)
**Reference Genome and Annotation: Rnor_6.0 assembly from Ensembl database (release 99)**
```{r}
################################################################################
#####     Annotate Genes by Chromosome       ###################################
################################################################################

### Determine which control samples are male and female
# Get the list of genes on the W chromosome

# Construct your own personal galgal5 reference genome annotation
# Construct from gtf file from Ensembl (same file used in mapping)
ens_gtf = paste0(WD,'/Data/library/Rattus_norvegicus.Rnor_6.0.99.gtf')

Rn_TxDb <- makeTxDbFromGFF(ens_gtf,
                           format=c("gtf"),
                           dataSource="Ensembl_Rattus6_gtf",
                           organism="Rattus norvegicus",
                           taxonomyId=NA,
                           circ_seqs=DEFAULT_CIRC_SEQS,
                           chrominfo=NULL,
                           miRBaseBuild=NA,
                           metadata=NULL)

seqlevels(Rn_TxDb)[1:23]

# Extract genes as GRanges object, then names
X_genes_gr <- genes(Rn_TxDb, columns = "TXCHROM", filter = list(tx_chrom=c("X")))
# Collect ensembl gene ids for female specific genes
X_ens_id <- names(X_genes_gr)
# Examine the gene symbols
X_sym <- mapIds(org.Rn.eg.db, names(X_genes_gr), "SYMBOL", "ENSEMBL")
# Extract genes as GRanges object, then names
Y_genes_gr <- genes(Rn_TxDb, columns = "TXCHROM", filter = list(tx_chrom=c("Y")))
# Collect ensembl gene ids for female specific genes
Y_ens_id <- names(Y_genes_gr)
sex_ens_id <- c(X_ens_id,Y_ens_id)
# Examine the gene symbols
Y_sym <- mapIds(org.Rn.eg.db, names(Y_genes_gr), "SYMBOL", "ENSEMBL")
```


### PCA Visualization
```{r}
count_data = sford_2[, sort(colnames(sford_2))]
col_data = status[sort(rownames(status)), ]



design = ~1
title = paste0('Design: ',as.character(design))
# Create a DESeqDataSet Object
dds <- DESeqDataSetFromMatrix(countData = count_data,
                              colData = col_data,
                              design = design)


# Filter genes with average count of 10 or less. (this time I won't perform this)
# Reasoning from:
#citation("PROPER")
#dds
# keep <- rowMeans(counts(dds)) >= 10
# dds <- dds[keep,]

dds <- estimateSizeFactors(dds)
rs <- rowSums(counts(dds))
# Normalize the counts
rld <- vst(dds) #vst and rlog comparable with all samples
#rld <- rlog(dds, blind=FALSE)

# Extract matrix of normalized counts
norm_counts <- assay(rld)
counts <- as.data.frame(norm_counts)

# Annotate normalized counts
counts$ensembl <- rownames(counts)
counts$symbol <- mapIds(org.Rn.eg.db, counts$ensembl, "SYMBOL", "ENSEMBL")
counts$entrez <- mapIds(org.Rn.eg.db, counts$ensembl, "ENTREZID", "ENSEMBL")
counts$genename <- mapIds(org.Rn.eg.db, counts$ensembl, "GENENAME", "ENSEMBL")
counts$go <- mapIds(org.Rn.eg.db, counts$ensembl, "GO", "ENSEMBL")
counts$path <- mapIds(org.Rn.eg.db, counts$ensembl, "PATH", "ENSEMBL")


# Here we examine all samples by batch and tissue
# 
# first PCA with intGroup of tissue
DESeq2::plotPCA(rld, intgroup ="Tissue") +
  guides(color=guide_legend(title="Tissues"))
```


```{r}
################################################################################
#####     Normalization for Sequencing Depth      ##############################
################################################################################
hypothalamus = col_data %>% filter(Tissue == 'Hypothalamus') %>%
  select(vial_label) %>% unlist() %>% as.character()
hypothalamus_counts = count_data[,hypothalamus] %>% as.matrix()
# subset hypothalamus metadata
hypothalamus_cols = col_data %>% filter(Tissue == 'Hypothalamus')
row.names(hypothalamus_cols) = hypothalamus_cols$vial_label
# Sanity Check: Ensure that the metadata rownames are identical to count matrix column names
all(rownames(hypothalamus_cols) == colnames(hypothalamus_counts))

hypo_dds <- DESeqDataSetFromMatrix(countData = hypothalamus_counts,
                              colData = hypothalamus_cols,
                              design = design)

# remove genes that have avg counts less than 1 (reduced to maintain more sex-relate genes)
keep <- rowSums(counts(hypo_dds))/ncol(hypo_dds) > 1
hypo_dds <- hypo_dds[keep,]

# see how gene counts distributed
plot(assay(hypo_dds)[,3:4], cex=.3)
# DESeq2 uses median of ratios to normalize sequencing depth and RNA composition (not for within samaple comparison)
hypo_dds = estimateSizeFactors(hypo_dds)
# then view size factors:
plot(sizeFactors(hypo_dds), colSums(counts(hypo_dds)), cex=.3)
abline(lm(colSums(counts(hypo_dds)) ~ sizeFactors(hypo_dds) + 0))
normalized_counts = counts(hypo_dds, normalized = T)
#summary(normalized_counts)
summary(sizeFactors(hypo_dds))
```
### Transformation

```{r}
# Comparison 3 transformation methods: log2 norm, vst and rlog
log2_norm_counts = log2(normalized_counts + 1)
vst_norm_counts = vst(hypo_dds)
rlog_norm_counts = rlog(hypo_dds)
par(mfrow=c(1,2))
boxplot(log2_norm_counts, ylim = c(0,20), cex = 0.2, main = "log2 transformation")
boxplot(assay(vst_norm_counts), ylim = c(0,20), cex = 0.2, main = "vst transformation")
boxplot(log2_norm_counts, ylim = c(0,20), cex = 0.2, main = "log2 transformation")
boxplot(assay(rlog_norm_counts), ylim = c(0,20), cex = 0.2, main = "rlog transformation")
```
* VST uses a subset of genes and divides each column of the counts matrix by size factors. It shrinked  
the counts matrix range (to about 6 to 20) more than the other 2 methods.
* Log2 and rlog are pretty much similar to each other.
```{r}
plot(assay(vst_norm_counts)[,3:4],
     xlab = "sample 1 VST gene counts",
     ylab = "sample 2 VST gene counts",
     cex = .3)
plot(assay(rlog_norm_counts)[,3:4],
     xlab = "sample 1 rlog gene counts",
     ylab = "sample 2 rlog gene counts",
     cex = .3)
plot(log2_norm_counts[,3:4],
     xlab = "sample 1 log2 gene counts",
     ylab = "sample 2 log2 gene counts",
     cex = .3)
```
* Rlog (regularized log) transformation showed the most normal distributed variance among all genes (from  
low counts to high counts), so here I continue on with rlog normalized S4.

### PCA by tissue == Hypothalamus
```{r}
counts = as.data.frame(assay(rlog_norm_counts))
# Annotate normalized counts
counts$ensembl <- rownames(counts)
counts$symbol <- mapIds(org.Rn.eg.db, counts$ensembl, "SYMBOL", "ENSEMBL")
counts$entrez <- mapIds(org.Rn.eg.db, counts$ensembl, "ENTREZID", "ENSEMBL")
counts$genename <- mapIds(org.Rn.eg.db, counts$ensembl, "GENENAME", "ENSEMBL")
counts$go <- mapIds(org.Rn.eg.db, counts$ensembl, "GO", "ENSEMBL")
counts$path <- mapIds(org.Rn.eg.db, counts$ensembl, "PATH", "ENSEMBL")

# Here we examine all samples by batch and tissue
# 
# first PCA with intGroup of tissue
DESeq2::plotPCA(rlog_norm_counts, intgroup ="animal.registration.sex") +
  guides(color=guide_legend(title="sex"))


pcaData <- DESeq2::plotPCA(rlog_norm_counts, intgroup=c("animal.key.anirandgroup","animal.registration.sex"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=animal.key.anirandgroup, shape = animal.registration.sex)) +
  geom_point(size=3) +
  #geom_text(aes(label=Tissue),hjust=0, vjust=0) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  #coord_fixed() +
  ggtitle("Hypothalamus Samples Across Groups")


```

```{r}
# Variables of interest
male_hypothalamus <- (hypothalamus_cols %>% 
                        filter(animal.registration.sex == 'Male'))$vial_label %>% 
        as.character()
female_hypothalamus <- (hypothalamus_cols %>% 
                          filter(animal.registration.sex == 'Female'))$vial_label %>% 
        as.character()
ref_hypothalamus <- (hypothalamus_cols %>%
                       filter(is.na(animal.registration.sex)))$vial_label %>% 
        as.character()
hypothalamus <- c(male_hypothalamus,female_hypothalamus,ref_hypothalamus)
Y_genes <- Y_ens_id[Y_ens_id %in% row.names(assay(rlog_norm_counts))]
X_genes <- X_ens_id[X_ens_id %in% row.names(assay(rlog_norm_counts))]
sex <- hypothalamus_cols[hypothalamus,"animal.registration.sex"]
group <- hypothalamus_cols[hypothalamus,"animal.key.anirandgroup"]


# t-test on expression associated with sex
tt = rowttests(hypothalamus_counts[, hypothalamus], sex)
par(myfrow = c(1,2))
# Histogram of p values associated with ttest
hist(tt$p.value,main="",ylim=c(0,1300), breaks = 100)
plot(tt$dm,-log10(tt$p.value))
points(tt[X_genes,]$dm,-log10(tt[X_genes,]$p.value),col=3,pch=16)
points(tt[Y_genes,]$dm,-log10(tt[Y_genes,]$p.value),col=2,pch=16, xlab="Effect size",ylab="-log10(p-value)")
legend("bottomright",c("X","Y"),col=3:2,pch=16)
p <- tt$p.value
qvals <- qvalue(tt$p.value)$qvalue
index <- which(qvals<=0.05)
abline(h=-log10(max(tt$p.value[index])))
```

### PCA after removing sex-differentially expressed genes 
#### (have I removed sex-related vairance?)
```{r}
# first try removing all sex-significant genes
hypothalamus_counts_rm_sig = hypothalamus_counts[!(rownames(hypothalamus_counts)) %in% rownames(hypothalamus_counts[index,hypothalamus]),]

# then try removing only X or Y related sex-significant genes
# here's a new index
sex_index = c(index[which(rownames(hypothalamus_counts[index,hypothalamus]) %in% X_genes)], index[which(rownames(hypothalamus_counts[index,hypothalamus]) %in% Y_genes)])

hypothalamus_counts_rm_sex = hypothalamus_counts[!(rownames(hypothalamus_counts)) %in% rownames(hypothalamus_counts[sex_index,hypothalamus]),]

# build dds with those 2 new count matrix
hypo_dds_rm_sig <- DESeqDataSetFromMatrix(countData = hypothalamus_counts_rm_sig,
                              colData = hypothalamus_cols,
                              design = design)

# remove genes that have avg counts less than 1 (reduced to maintain more sex-relate genes)
keep <- rowSums(counts(hypo_dds_rm_sig))/ncol(hypo_dds_rm_sig) > 1
hypo_dds_rm_sig <- hypo_dds_rm_sig[keep,]
hypo_dds_rm_sig = estimateSizeFactors(hypo_dds_rm_sig)
rlog_norm_hypo_dds_rm_sig = rlog(hypo_dds_rm_sig)


DESeq2::plotPCA(rlog_norm_hypo_dds_rm_sig, intgroup ="animal.registration.sex") +
  guides(color=guide_legend(title="sex"))


pcaData <- DESeq2::plotPCA(rlog_norm_hypo_dds_rm_sig, intgroup=c("animal.key.anirandgroup","animal.registration.sex"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=animal.key.anirandgroup, shape = animal.registration.sex)) +
  geom_point(size=3) +
  #geom_text(aes(label=Tissue),hjust=0, vjust=0) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  #coord_fixed() +
  ggtitle("Hypothalamus Samples With 22 Genes Removed")
```
```{r}
hypo_dds_rm_sex <- DESeqDataSetFromMatrix(countData = hypothalamus_counts_rm_sex,
                              colData = hypothalamus_cols,
                              design = design)

# remove genes that have avg counts less than 1 (reduced to maintain more sex-relate genes)
keep <- rowSums(counts(hypo_dds_rm_sex))/ncol(hypo_dds_rm_sex) > 1
hypo_dds_rm_sex <- hypo_dds_rm_sex[keep,]
hypo_dds_rm_sex = estimateSizeFactors(hypo_dds_rm_sex)
rlog_norm_hypo_dds_rm_sex = rlog(hypo_dds_rm_sex)


DESeq2::plotPCA(rlog_norm_hypo_dds_rm_sex, intgroup ="animal.registration.sex") +
  guides(color=guide_legend(title="sex"))


pcaData <- DESeq2::plotPCA(rlog_norm_hypo_dds_rm_sex, intgroup=c("animal.key.anirandgroup","animal.registration.sex"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=animal.registration.sex)) +
  geom_point(size=3) +
  #geom_text(aes(label=Tissue),hjust=0, vjust=0) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  #coord_fixed() +
  ggtitle("Hypothalamus Samples with 16 Genes Removed")
```

## Defferential gene expression
### Modeling raw counts with normalization

now perform *differential gene expression* on the counts, try to find genes with dfference in counts across groups due to the condition of interest rises above the biological and technical variance we observe

first prepare for col data and count data
```{r}
# variables in design formula cannot contain NA, so I have to remove ref samples here
hypothalamus_counts_rm_sex = hypothalamus_counts_rm_sex[,-which(is.na(col_data$animal.key.anirandgroup))]
hypothalamus_cols = hypothalamus_cols[-which(is.na(col_data$animal.key.anirandgroup)),]

# group all samples by acute exercise or not c(0,1)
for (i in 1:nrow(hypothalamus_cols)){
  if (hypothalamus_cols[i,]$animal.key.anirandgroup %in% c("Exercise - 0.5 hr", "Exercise - 1 hr", "Exercise - IPE")){
    hypothalamus_cols[i,]$acute_ex = 1
  }else{
    hypothalamus_cols[i,]$acute_ex = 0
  }
}
hypothalamus_cols$acute_ex = as.factor(hypothalamus_cols$acute_ex)

# one more step to remove batch effect between male and female
f_counts = hypothalamus_counts_rm_sex[, female_hypothalamus[ which(female_hypothalamus %in% colnames(hypothalamus_counts_rm_sex))]]
m_counts = hypothalamus_counts_rm_sex[, male_hypothalamus[ which(male_hypothalamus %in% colnames(hypothalamus_counts_rm_sex))]]
f_median = rowMedians(f_counts)
m_median = rowMedians(m_counts)
f_var = matrixStats::rowVars(f_counts)
m_var = matrixStats::rowVars(m_counts)

# for lots of genes, there's 0 expression in all samples which results in a variance of 0
# they will be removed after building dds object, so I'll just keep them here

new_var = pmin(f_var, m_var)
f_var_scale = ifelse(f_var == 0, 0, sqrt(new_var / f_var))
m_var_scale = ifelse(m_var == 0, 0, sqrt(new_var / m_var))
for (i in 1:nrow(hypothalamus_counts_rm_sex)){
  f_counts[i] = f_counts[i] + max(f_median[i], m_median[i]) - f_median[i]
  m_counts[i] = m_counts[i] + max(f_median[i], m_median[i]) - m_median[i]
  # m_counts[i] = ifelse(m_var_scale == 0, m_counts[i], m_var_scale[i] * m_counts[i] - (m_var_scale[i] - 1) * mean(m_counts[i])) %>% as.integer()
  # f_counts[i] = ifelse(f_var_scale == 0, f_counts[i], f_var_scale[i] * f_counts[i] - (f_var_scale[i] - 1) * mean(f_counts[i])) %>% as.integer()
}

hypothalamus_counts_scaled = cbind(f_counts, m_counts)
hypothalamus_cols_scaled = hypothalamus_cols[c(female_hypothalamus[ which(female_hypothalamus %in% colnames(hypothalamus_counts_rm_sex))], male_hypothalamus[ which(male_hypothalamus %in% colnames(hypothalamus_counts_rm_sex))]),]
```

```{r}
# Create a DESeqDataSet Object
dds <- DESeqDataSetFromMatrix(countData = hypothalamus_counts_scaled,
                              colData = hypothalamus_cols_scaled,
                              design = ~ animal.registration.sex + acute_ex)
# based on the vignette of DESeq2: acute_ex (the last variable in design) effect represents the overall effect controlling for differences due to sex

# Filter genes with average count of 10 or less. (this time I won't perform this)
# Reasoning from:
#citation("PROPER")
#dds
keep <- rowMeans(counts(dds)) >= 1
dds <- dds[keep,]

dds_1 <- estimateSizeFactors(dds)
# Normalize the counts
rld <- rlog(dds_1) #vst and rlog comparable with all samples
#rld <- rlog(dds, blind=FALSE)

# Extract matrix of normalized counts
norm_counts <- assay(rld)
counts <- as.data.frame(norm_counts)

# Annotate normalized counts
counts$ensembl <- rownames(counts)
counts$symbol <- mapIds(org.Rn.eg.db, counts$ensembl, "SYMBOL", "ENSEMBL")
counts$entrez <- mapIds(org.Rn.eg.db, counts$ensembl, "ENTREZID", "ENSEMBL")
counts$genename <- mapIds(org.Rn.eg.db, counts$ensembl, "GENENAME", "ENSEMBL")
counts$go <- mapIds(org.Rn.eg.db, counts$ensembl, "GO", "ENSEMBL")
counts$path <- mapIds(org.Rn.eg.db, counts$ensembl, "PATH", "ENSEMBL")

pcaData <- DESeq2::plotPCA(rld, intgroup=c("animal.key.anirandgroup","animal.registration.sex"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=animal.key.anirandgroup, shape = animal.registration.sex)) +
  geom_point(size=3) +
  #geom_text(aes(label=Tissue),hjust=0, vjust=0) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  #coord_fixed() +
  ggtitle("Hypothalamus Samples with sex median-centered") 
```
### to Compare whether I have removed the sex-related variance
* subset by sex == female
* PCA show the data structure

```{r}
hypothalamus_cols_f = hypothalamus_cols[female_hypothalamus,]
hypothalamus_counts_f = hypothalamus_counts[,female_hypothalamus]
# Create a DESeqDataSet Object
dds_f <- DESeqDataSetFromMatrix(countData = hypothalamus_counts_f,
                              colData = hypothalamus_cols_f,
                              design = ~ 1)


# Filter genes with average count of 10 or less. (this time I won't perform this)
# Reasoning from:
#citation("PROPER")
#dds
keep <- rowMeans(counts(dds_f)) >= 1
dds_f <- dds_f[keep,]

dds_f <- estimateSizeFactors(dds_f)
# Normalize the counts
rld <- rlog(dds_f) #vst and rlog comparable 

pcaData <- DESeq2::plotPCA(rld, intgroup=c("animal.key.anirandgroup","acute_ex"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=animal.key.anirandgroup, shape = acute_ex)) +
  geom_point(size=3) +
  #geom_text(aes(label=Tissue),hjust=0, vjust=0) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  #coord_fixed() +
  ggtitle("Hypothalamus Samples (Female only)")
```






Then run the DESeq2 model:
```{r}
dds <- DESeq2::DESeq(dds)
res = results(dds)
head(res)

```
**for myself reference**
Another way to look at the difference is that a p-value of 0.05 implies that 5% of all tests  
will result in false positives. An FDR adjusted p-value (or q-value) of 0.05 implies that 5%   
of significant tests will result in false positives. The latter will result in fewer false  
positives.  
```{r}
# table(res$padj < 0.1)
# summary(res)
#which()
res2 <- results(dds, alpha=0.05)
table(res2$padj < 0.05)

# a sorted results table:
resSort <- res2[order(res2$padj),]

gene_ls = counts[rownames(res2[which(res2$padj <= 0.05),]),]$symbol
gene_ls = gene_ls[!is.na(gene_ls)]
write.table(gene_ls, file = paste0(WD, "/Data/intermediate/20200423-hypothalamus-DE-gene-list_Zhang.txt"), sep = "\t", col.names = F, row.names = F, quote = FALSE)

```
### Visualizing results
```{r}
plotMA(res2, ylim=c(-4,4), alpha = 0.05)
hist(res2$pvalue[res2$baseMean > 1], 
     col="grey", border="white", xlab="pvalue", ylab="frequency", main="")
```
Examine the counts for the top gene, sorting by FDR:
```{r}
#plotCounts(dds, gene=which.min(res2$padj), intgroup="animal.key.anirandgroup")
data <- plotCounts(dds, gene=which.min(res2$padj), intgroup="animal.key.anirandgroup", returnData=TRUE)
ggplot(data, aes(x=animal.key.anirandgroup, y=count, col = hypothalamus_cols$hour_death)) +
  geom_point(position=position_jitter(width=.1,height=0)) +
  #stat_smooth() + 
  scale_y_log10() +
  theme(axis.text.x = element_text(angle = 45))

#which gene it is
counts[which.min(res2$padj),c("symbol","genename")]
```
Draw heatmap of the top genes:

```{r}
topgenes <- head(rownames(resSort),20)
mat <- assay(rld)[topgenes,]

# standarized the count matrix
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds)$animal.key.anirandgroup)
rownames(df) = colnames(mat)
pheatmap(mat, annotation_col=df)
```
Generally, exercise-IPE and exercise-0.5hr are clustered together, apart from other groups.  


As we talked in meeting, the control-7 group and exercise-7 group are well matched  
So I try to find genes that are significantly differently expressed in those 2 groups
```{r}


res3 <- results(dds, alpha=0.05, contrast = c("animal.key.anirandgroup", "Control - 7 hr", "Exercise - 7 hr"))
table(res3$padj < 0.05)
hist(res3$pvalue[res3$baseMean > 1], 
     col="grey", border="white", xlab="pvalue", ylab="frequency", main="")
data <- plotCounts(dds, gene=which.min(res3$padj), intgroup="animal.key.anirandgroup", returnData=TRUE)
ggplot(data, aes(x=animal.key.anirandgroup, y=count, col = col_data$hour_death)) +
  geom_point(position=position_jitter(width=.1,height=0)) +
  #scale_y_log10() +
  theme(axis.text.x = element_text(angle = 45))

counts[which.min(res3$padj),c("symbol","genename")]
```


Also draw a heatmap for the top DE genes in the control-7 group vs the exercise-7 group:
```{r}
resSort_7 <- res3[order(res3$padj),]
topgenes_7 <- head(rownames(resSort_7),20)
mat <- assay(rld)[topgenes_7,]

# standarized the count matrix
mat <- mat - rowMeans(mat)
pheatmap(mat, annotation_col=df)
```

